minvws/nl-kat-coordination
# Scheduler rate limit functionality

### Changes

This PR adds rate-limiting capabilities to the scheduler for tasks. Boefjes that are subject to rate limiting (defined in the boefje manifest) are being tracked by an in-memory rate limiter, leveraging the package [limits](https://limits.readthedocs.io/en/stable/index.html). When a task is subject to a rate-limit it will get the `DELAYED` status and will be postponed until it is allowed again. A thread will check for delayed tasks and put them on the queue when they're ready.

This will expect some changes in the katalogus mainly exposing a grouping/namespacing on which to rate limit and the [parseable rate limit](https://limits.readthedocs.io/en/stable/quickstart.html#rate-limit-string-notation).

Example:

- grouping/namepace: `org-1/external-api-key`
- rate limit: `10/hour;100/day;2000 per year`

### Issue link
Closes #1317 

### Processess

For a boefje scheduler when a task gets created it will get check whether the boefje is rate limited or not:

```mermaid
flowchart TD
    A[Task Creation] --> B{is_task_rate_limited}
    B -->|Yes| C[set Task to status <br>DELAYED]
    B ---->|No| E[End]
```

Checking if a task is rate limited:

```mermaid
graph TD
    A[Start] --> B{rate_limit is None?}
    B -- Yes --> C[Return False]
    B -- No --> D{Try to parse rate_limit}
    D --> E{parsed_rate_limit is None?}
    E -- Yes --> F[Raise ValueError]
    E -- No --> G[Lock rate_limiter]
    G --> H[Get identifier_template]
    H --> I[Render identifier from template]
    I --> J[Test rate_limiter with identifier]
    J --> K{can_consume?}
    K -- No --> L[Return True]
    K -- Yes --> M{hit?}
    M -- Yes --> N[Hit rate_limiter]
    N --> O[Return False]
    M -- No --> O
    D --> P[Caught ValueError]
    P --> Q[Log warning]
    Q --> R[Raise exc]
```

Within the boefje scheduler a new process (`push_tasks_for_delayed_tasks`) is running next to:

* check for scan profile changes
* check for enabled boefjes
* check for rescheduling (random ooi endpoint octopoes)

This `push_tasks_for_delayed_tasks` method gets all tasks that have the status of `DELAYED`. And will be checked whether or not task can be pushed onto the queue.
